File: lib/protocol/channel_client.dart
import 'dart:async';
import 'dart:typed_data';

import 'constants.dart';
import 'value_codec.dart';

class ChannelRpcError implements Exception {
  final String message;
  ChannelRpcError(this.message);

  @override
  String toString() => 'ChannelRpcError: $message';
}

/// Channel RPC over the bridge: request header
/// `[reqType, reqId, channelName, name]` + argument value.
class ChannelClient {
  final void Function(Uint8List body) sendBody;
  final void Function(String line)? onLog;

  int _lastRequestId = 0;
  final _ready = Completer<void>();
  final _handlers = <int, void Function(int type, Object? data)>{};
  final _pending = <int, Completer<dynamic>>{};
  bool _disposed = false;

  ChannelClient({required this.sendBody, this.onLog});

  Future<void> get ready => _ready.future;

  void handleMessage(Uint8List body) {
    try {
      final reader = ValueReader(body);
      final header = reader.readValue();
      if (header is! List || header.isEmpty || header[0] is! num) return;
      final type = (header[0] as num).toInt();
      if (type == ipcResInitialize) {
        onLog?.call('[ipc] initialized');
        if (!_ready.isCompleted) _ready.complete();
        return;
      }
      if (header.length < 2 || header[1] is! num) return;
      final id = (header[1] as num).toInt();
      final data = reader.readValue();
      _handlers[id]?.call(type, data);
    } on Object catch (e) {
      onLog?.call('[ipc] invalid frame: $e');
    }
  }

  Future<dynamic> call(
    String channel,
    String method,
    List<Object?> args, {
    Duration timeout = const Duration(seconds: 30),
  }) async {
    // 桥已拆：调用方还在排队的收尾动作要立刻知道，别干等满超时。
    if (_disposed) throw StateError('channel disposed');
    await ready.timeout(
      const Duration(seconds: 30),
      onTimeout: () {
        throw TimeoutException('channel init timeout');
      },
    );
    final id = _lastRequestId++;
    final completer = Completer<dynamic>();
    _pending[id] = completer;
    _handlers[id] = (type, data) {
      switch (type) {
        case ipcResPromiseSuccess:
          _handlers.remove(id);
          _pending.remove(id);
          completer.complete(data);
        case ipcResPromiseError:
          _handlers.remove(id);
          _pending.remove(id);
          final message = data is Map
              ? (data['message'] ?? data).toString()
              : '$data';
          completer.completeError(ChannelRpcError(message));
        case ipcResPromiseErrorObj:
          _handlers.remove(id);
          _pending.remove(id);
          completer.completeError(ChannelRpcError('$data'));
      }
    };
    onLog?.call('[ipc] call $channel.$method id=$id');
    _send(ipcReqPromise, id, channel, method, args);
    return completer.future.timeout(
      timeout,
      onTimeout: () {
        _handlers.remove(id);
        _pending.remove(id);
        throw TimeoutException('$channel.$method timed out', timeout);
      },
    );
  }

  /// Subscribe to a channel event; returns a cancel function.
  void Function() addEventListener(
    String channel,
    String event,
    void Function(Object? event) onEvent, {
    Object? arg,
  }) {
    final id = _lastRequestId++;
    var sent = false;
    var cancelled = false;
    _handlers[id] = (type, data) {
      if (type == ipcResEventFire) onEvent(data);
    };
    ready.then((_) {
      if (cancelled) return;
      sent = true;
      onLog?.call('[ipc] listen $channel.$event id=$id');
      _send(ipcReqEventListen, id, channel, event, arg);
    });
    return () {
      cancelled = true;
      _handlers.remove(id);
      if (sent) _send(ipcReqEventDispose, id, channel, event, null);
    };
  }

  void _send(int reqType, int id, String channel, String name, Object? arg) {
    final writer = ValueWriter();
    writer.writeValue([reqType, id, channel, name]);
    writer.writeValue(arg);
    sendBody(writer.take());
  }

  /// 桥栈切换时在途调用立即报错，别让调用方干等满超时（表现为"点了没反应"）。
  /// 置 `_disposed` 后**后续**调用也会立刻报错——拆栈期间还有收尾动作在排队
  /// 时，它们不该各自再等一次完整超时。
  void dispose() {
    _disposed = true;
    for (final completer in _pending.values) {
      if (!completer.isCompleted) {
        completer.completeError(StateError('channel disposed'));
      }
    }
    _pending.clear();
    _handlers.clear();
  }
}

File: web/src/protocol/channelClient.ts
/// Channel RPC over bridge — TS 移植自 lib/protocol/channel_client.dart。
/// 请求头 `[reqType, reqId, channelName, name]` + 参数 value，二进制走 ValueCodec。

import {
  IPC_REQ_PROMISE,
  IPC_REQ_EVENT_LISTEN,
  IPC_REQ_EVENT_DISPOSE,
  IPC_RES_INITIALIZE,
  IPC_RES_PROMISE_SUCCESS,
  IPC_RES_PROMISE_ERROR,
  IPC_RES_PROMISE_ERROR_OBJ,
  IPC_RES_EVENT_FIRE,
} from './constants'
import { ValueReader, ValueWriter } from './valueCodec'

export class ChannelRpcError extends Error {
  constructor(message: string) {
    super(`ChannelRpcError: ${message}`)
    this.name = 'ChannelRpcError'
  }
}

type ResHandler = (type: number, data: unknown) => void

export class ChannelClient {
  private lastRequestId = 0
  private readyResolve: (() => void) | null = null
  private readyReady = false
  private handlers = new Map<number, ResHandler>()
  private pending = new Map<number, { resolve: (v: unknown) => void; reject: (e: unknown) => void }>()
  private disposed = false

  constructor(
    private sendBody: (body: Uint8Array) => void,
    private onLog?: (line: string) => void,
  ) {}

  /** 立刻可 await 的 ready promise（对齐 Dart 的 Completer）。 */
  get ready(): Promise<void> {
    if (this.readyReady) return Promise.resolve()
    return new Promise((resolve) => {
      this.readyResolve = resolve
    })
  }

  handleMessage(body: Uint8Array): void {
    try {
      const reader = new ValueReader(body)
      const header = reader.readValue()
      if (!Array.isArray(header) || header.length === 0 || typeof header[0] !== 'number') return
      const type = header[0]
      if (type === IPC_RES_INITIALIZE) {
        this.onLog?.('[ipc] initialized')
        if (!this.readyReady) {
          this.readyReady = true
          this.readyResolve?.()
        }
        return
      }
      if (header.length < 2 || typeof header[1] !== 'number') return
      const id = header[1]
      const data = reader.readValue()
      this.handlers.get(id)?.(type, data)
    } catch (e) {
      this.onLog?.(`[ipc] invalid frame: ${e}`)
    }
  }

  async call(
    channel: string,
    method: string,
    args: unknown[],
    timeoutMs = 30_000,
  ): Promise<unknown> {
    if (this.disposed) throw new Error('channel disposed')
    await this.waitForReady()
    const id = this.lastRequestId++
    const pending = this.pending.get.bind(this.pending)
    void pending
    return new Promise((resolve, reject) => {
      const timer = setTimeout(() => {
        this.handlers.delete(id)
        this.pending.delete(id)
        reject(new Error(`${channel}.${method} timed out`))
      }, timeoutMs)
      const settle = (fn: () => void) => {
        clearTimeout(timer)
        fn()
      }
      this.handlers.set(id, (type, data) => {
        switch (type) {
          case IPC_RES_PROMISE_SUCCESS:
            settle(() => {
              this.handlers.delete(id)
              this.pending.delete(id)
              resolve(data)
            })
            break
          case IPC_RES_PROMISE_ERROR:
            settle(() => {
              this.handlers.delete(id)
              this.pending.delete(id)
              const message =
                data && typeof data === 'object' && 'message' in data
                  ? String((data as Record<string, unknown>)['message'])
                  : String(data)
              reject(new ChannelRpcError(message))
            })
            break
          case IPC_RES_PROMISE_ERROR_OBJ:
            settle(() => {
              this.handlers.delete(id)
              this.pending.delete(id)
              reject(new ChannelRpcError(String(data)))
            })
            break
        }
      })
      this.pending.set(id, { resolve, reject })
      this.onLog?.(`[ipc] call ${channel}.${method} id=${id}`)
      this.send(IPC_REQ_PROMISE, id, channel, method, args)
    })
  }

  private waitForReady(): Promise<void> {
    if (this.readyReady) return Promise.resolve()
    return Promise.race([
      this.ready,
      new Promise<never>((_, reject) =>
        setTimeout(() => reject(new Error('channel init timeout')), 30_000),
      ),
    ])
  }

  /** 订阅通道事件；返回取消函数。 */
  addEventListener(
    channel: string,
    event: string,
    onEvent: (data: unknown) => void,
    arg?: unknown,
  ): () => void {
    const id = this.lastRequestId++
    let sent = false
    let cancelled = false
    this.handlers.set(id, (type, data) => {
      if (type === IPC_RES_EVENT_FIRE) onEvent(data)
    })
    void this.ready.then(() => {
      if (cancelled) return
      sent = true
      this.onLog?.(`[ipc] listen ${channel}.${event} id=${id}`)
      this.send(IPC_REQ_EVENT_LISTEN, id, channel, event, arg)
    })
    return () => {
      cancelled = true
      this.handlers.delete(id)
      if (sent) this.send(IPC_REQ_EVENT_DISPOSE, id, channel, event, null)
    }
  }

  private send(reqType: number, id: number, channel: string, name: string, arg: unknown): void {
    const writer = new ValueWriter()
    writer.writeValue([reqType, id, channel, name])
    writer.writeValue(arg)
    this.sendBody(writer.take())
  }

  /** 桥栈切换时在途调用立刻报错。 */
  dispose(): void {
    this.disposed = true
    for (const p of this.pending.values()) p.reject(new Error('channel disposed'))
    this.pending.clear()
    this.handlers.clear()
  }
}
